mtd: rawnand: Add Phytium NAND flash controller support

studyhu opened the pull request:
This patch implements a device driver of NAND flash controller for both Phytium PCIe chipset and SoCs. Since interrupt flags depends on the bus implementation, e.g., PCI vs platform, we extract the interrupt register function from NFC driver core.

Avenger-285714 commented on drivers/mtd/nand/raw/phytium_nand.c:
```diff
@@ -0,0 +1,2089 @@
+
+	for (i = 0; i < cmd_num; i++) {
+		memcpy((u8 *)nfc->dsp_addr + i*PHYTIUM_NFC_DSP_SIZE,
+			(u8 *)nfc_op, PHYTIUM_NFC_DSP_SIZE);
```
建议判断一下cmd_num有没有可能导致越界，下同

studyhu replied:
在phytium_nfc_init_dma函数中，给nfc->dsp_addr 申请PAGE_SIZE的内存大小，然后调用phytium_nfc_send_cmd2函数时，传入的参数cmd_num都是2，PAGE_SIZE要大于i*PHYTIUM_NFC_DSP_SIZE，不会越界。

opsiff commented:
/ok-to-test

Avenger-285714 commented on drivers/mtd/nand/raw/phytium_nand.c:
```diff
@@ -0,0 +1,2089 @@
+}
+
+static int phytium_nand_get_ecc_total(struct mtd_info *mtd,
+				 struct nand_ecc_ctrl *ecc)
```
代码对齐

Avenger-285714 commented on drivers/mtd/nand/raw/phytium_nand.c:
```diff
@@ -0,0 +1,2089 @@
+
+	ret = phytium_nfc_wait_op(chip, nfc_op->rdy_timeout_ms);
+	if (ret)
+		return ret;
```
kfree(nfc_op);

studyhu replied:
已修改

Avenger-285714 commented on drivers/mtd/nand/raw/phytium_nand.c:
```diff
@@ -0,0 +1,2089 @@
+	sdr = nand_get_sdr_timings(nand_get_interface_config(&phytium_nand->chip));
+	ecc_offset = phytium_nand->ecc.offset;
+
+	nfc_op = kzalloc(2 * sizeof(struct phytium_nfc_op), GFP_KERNEL);
```
nfc_op没释放

studyhu replied:
已修改

Avenger-285714 commented on drivers/mtd/nand/raw/phytium_nand.c:
```diff
@@ -0,0 +1,2089 @@
+	/* Configure the DMA  */
+	phytium_nfc_init_dma(nfc);
+
+
```
空行太多

Commit:
mtd: rawnand: Add Phytium NAND flash controller support

This patch implements a device driver of NAND flash controller
for both Phytium PCIe chipset and SoCs. Since interrupt flags
depends on the bus implementation, e.g., PCI vs platform,
we extract the interrupt register function from NFC driver core.

Signed-off-by: Zhu Mingshuai <zhumingshuai@phytium.com.cn>
Signed-off-by: Zhou Yulin <zhouyulin1283@phytium.com.cn>
Signed-off-by: Chen Baozi <chenbaozi@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Signed-off-by: Hu Yeqing <huyeqing2074@phytium.com.cn>
